external-crates/move/crates/move-analyzer/src/symbols.rs

```diff
@@ -1831,10 +1842,14 @@ impl<'a> TypingSymbolicator<'a> {
                 &name_start,
                 &pos.file_hash(),
             );
-            let mod_sym = self.mod_outer_defs.get(&mod_ident).unwrap();
-            let fun_def = mod_sym.functions.get(name).unwrap();
-            let on_hover = fun_def.on_hover.clone();
-            let fun_type_def = on_hover_to_type_def_loc(self.mod_outer_defs, &on_hover);
+            let fun_info = self
+                .def_info
+                .get(&DefLoc {
+                    fhash: pos.file_hash(),
+                    start: name_start,
+                })
+                .unwrap();
```

Comment:
Uninformed question,  I've noticed there tends to be a lot of `unwrap`s and we're obviously (hopefully) preserving some invariant on these.

What's the general thing / convention I should know to convince myself this `unwrap` and others are safe (in this particular code)?

Reply:
Good point! While we are  building symbols information, certain things happen before others. In particular, we start by recording information about all top-level definitions in all modules. If later during the analysis, I refer to those, I know that this information is already part of the metadata. In this particular case, `pos` comes from previously processed top-level defs (and `def_info` is populated at the same time these top-level defs are processed) so unwrap is safe.

Smilarly (in other places), before a use of identifier is created, the algorithm checks if its definition is there (and only creates this use if it is). Consequently, we can rely on definition to be there if we grab it from the use (represented as `UseDef`).

Conversely, I am more careful in situations when I am not sure if something does or does not exist. For example, if I can't find a definition of a use, I will not record this use instead of crashing the server.

Reply:
Great, thanks for those details!